src/orthonormal_bases.jl

```diff
@@ -0,0 +1,186 @@
+end
+
+function represent_in_basis(M::Manifold, x, v, B::PrecomputedOrthonormalBasis)
+    return map(vb -> real(inner(M, x, v, vb)), B.vectors)
+end
+
+"""
+    inverse_represent_in_basis(M::Manifold, x, v, B::AbstractBasis)
+
+Convert a one-dimensional vector of coefficients in a basis `B` of
+the tangent space at `x` on manifold `M` to a tangent vector `v` at `x`.
+
+Depending on the basis, `x` may not directly represent a point on the manifold.
+For example if a basis transported along a curve is used, `x` may be the coordinate
+along the curve.
+
+See also: [`represent_in_basis`](@ref), [`basis`](@ref)
+"""
+function inverse_represent_in_basis(M::Manifold, x, v, B::AbstractBasis)
+    error("inverse_represent_in_basis not implemented for manifold of type $(typeof(M)) a point of type $(typeof(x)), tangent vector of type $(typeof(v)) and basis of type $(typeof(B)).")
+end
+
+function inverse_represent_in_basis(M::Manifold, x, v, B::PrecomputedOrthonormalBasis)
+    # quite convoluted but:
+    #  1) preserves the correct `eltype`
+    #  2) guarantees a reasonable array type `vout`
+    #     (for example scalar * `SizedArray` is an `SArray`)
+    vt = v[1] .* B.vectors[1]
+    vout = similar(B.vectors[1], eltype(vt))
+    copyto!(vout, vt)
+    for i in 2:length(v)
+        vout .+= v[i] .* B.vectors[i]
+    end
+    return vout
+end
+
+"""
+    basis(M::Manifold, x, B::AbstractBasis) -> AbstractBasis
+
+Compute the basis vectors of the tangent space at a point on manifold `M`
+represented by `x`.
+
+Returned object derives from [`AbstractBasis`](@ref) and has a field `.vectors`
+that stores tangent vectors.
+
+See also: [`represent_in_basis`](@ref), [`inverse_represent_in_basis`](@ref)
+"""
+function basis(M::Manifold, x, B::AbstractBasis)
+    error("basis not implemented for manifold of type $(typeof(M)) a point of type $(typeof(x)) and basis of type $(typeof(B)).")
+end
+
+basis(M::Manifold, x, B::AbstractPrecomputedOrthonormalBasis) = B
+
+function basis(M::ArrayManifold, x, B::AbstractPrecomputedOrthonormalBasis)
+    N = length(B)
+    M_dim = manifold_dimension(M)
+    N == M_dim || throw(ArgumentError("Incorrect number of basis vectors; expected: $M_dim, given: $N"))
+    for i in 1:N
+        vi_norm = norm(M, x, B.vectors[i])
+        isapprox(vi_norm, 1) || throw(ArgumentError("vector number $i is not normalized (norm = $vi_norm)"))
+        for j in i+1:N
+            dot_val = real(inner(M, x, B.vectors[i], B.vectors[j]))
+            isapprox(dot_val, 0; atol = eps(eltype(x))) || throw(ArgumentError("vectors number $i and $j are not orthonormal (inner product = $dot_val)"))
+        end
+    end
+    return B
+end
+
+function _euclidean_basis_vector(x, i)
+    y = zero(x)
+    y[i] = 1
+    return y
+end
+
+function basis(M::Manifold, x, B::ProjectedOrthonormalBasis{:svd})
```

Comment:
You could define a default `represent_in_basis` and `inverse_represent_in_basis` for `ProjectedOrthonormalBasis` like this:
```julia
represent_in_basis(M::Manifold, x, v, B::ProjectedOrthonormalBasis) = represent_in_basis(M, x, v, basis(M, x, B))
inverse_represent_in_basis(M::Manifold, x, v, B::ProjectedOrthonormalBasis) = inverse_represent_in_basis(M, x, v, basis(M, x, B))
```
with clear documentation that it's more efficient to precompute if one intends to reuse.

Reply:
I've been thinking about it but this default is so inefficient that I'd like to steer people away from doing this. People who don't want to store the basis on their side should use `ArbitraryOrthonormalBasis` (or some other efficient method we may have in the future).

It could be a good idea, though, to implement `ArbitraryOrthonormalBasis` using projected `basis` if we don't have a better idea.